python/openmlconverter/src/converter.py

```diff
@@ -0,0 +1,151 @@
+                        "source": f"#{{{distribution_source}/{feat['name']}}}",
+                    }
+                    for feat in openml_features
+                ],
+            }
+        ],
+    }
+    _remove_none_values(croissant)
+    for recordSet in croissant["recordSet"]:
+        _remove_none_values(recordSet)
+    return croissant
+
+
+def _get_field(json_dict, field: str, transform=None, required=False):
+    """
+    Convenience function. Get field and optionally perform the transformation. If field does not
+    exist, throw an error if the field is  required, otherwise return None.
+    """
+    if field not in json_dict:
+        if required:
+            raise ValueError(f"Required field {field} missing.")
+        return None
+    val = json_dict[field]
+    if transform is not None:
+        return transform(val)
+    return val
+
+
+def _person(name: str) -> dict | None:
+    if name is None or name == "":
+        return None
+    return {"@context": "https://schema.org", "@type": "Person", "name": name}
+
+
+def _distribution(name: str, url: str) -> dict | None:
+    if url is None:
+        return None
+
+    if url.endswith(".arff"):
+        type_ = "arff"
+        mimetype = "text/plain"  # No official arff mimetype exists
+    elif url.endswith(".pq"):
+        type_ = "parquet"
+        mimetype = "application/vnd.apache.parquet"  # Not an official mimetype yet
+        # see https://issues.apache.org/jira/browse/PARQUET-1889
+    else:
+        raise ValueError(f"Unrecognized file extension in url: {url}")
+    return {
+        "name": f"{name} ({type_})",
+        "@type": "FileObject",
+        "contentUrl": url,
+        "encodingFormat": mimetype
+        # TODO: add the md5 hash? But where? 'sha256' is in the Google Docs file, but not official?
+        # TODO: add sameAs (other distribution)?
+    }
+
+
+def _datatype(value: str, nominal_value: list[str] | None) -> str:
+    if nominal_value is not None:
+        if set(nominal_value) == {"TRUE", "FALSE"} or set(nominal_value) == {"0", "1"}:
+            return "Boolean"
+    d_type = {
+        "numeric": "Number",
+        "string": "Text",
+        "nominal": "Text",  # TODO: where to add the possible values?
+    }.get(value, None)
+    if d_type is None:
+        raise ValueError(f"Unknown datatype: {value}")
+    return d_type
+
+
+def _row_identifier(openml_features):
+    row_identifiers = [
+        f"#{{{f['name']}}}" for f in openml_features if f["is_row_identifier"] == "true"
+    ]
+    if len(row_identifiers) > 2:
+        return row_identifiers
+    elif len(row_identifiers) == 1:
+        return row_identifiers[0]
+    return None
+
+
+def _lenient_date_parser(value: str) -> datetime.date | datetime.datetime:
+    try:
+        dateutil.parser.parse(value)
+    except dateutil.parser.ParserError:
+        pass
+    if len(value) == 4 and (value.startswith("19") or value.startswith("20")):
```

Comment:
I would remove `and (value.startswith("19") or value.startswith("20")`: if we are in this function, then we already know that's supposed to be a date, no?
Maybe you'll have a dataset which an older book published before 1900?

Reply:
I understand your point, but personally I like to throw an exception in such weird cases. I'd like to look at a couple such examples. I'd suggest keeping the Exception and removing it after seeing an instance such as 1899 which is actually correct (and not a value such as 1234). I'm a bit afraid of the content of this field. It should be only dates, but the content format is not very strict.

What do you think?

Reply:
Works for me.